### PR TITLE
Migrating to protobuf-javalite

### PR DESCRIPTION
**Pull request by tomasz-ludek**
Based on https://github.com/bitcoinj/bitcoinj/pull/1938

I didn't disable tests in the `DeterministicKeyChainTest` failing by changed textual protobuf format since other tests were already broken and should be fixed altogether.

All protos were generated by _protoc-3.11.0-win64_
https://github.com/protocolbuffers/protobuf/releases/tag/v3.11.0
using `update-protobuf` maven profile:
```
mvn clean compile -DupdateProtobuf=true
```
`mvn clean compile -P update-protobuf` works as well

(on Windows the _protoc.exe_ has to be placed in project main dir)

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java [hunk 1]
```diff
@@ -786,7 +786,7 @@ public boolean isFollowing() {
             DeterministicKey key = (DeterministicKey) entry.getKey();
             Protos.Key.Builder proto = entry.getValue();
             proto.setType(Protos.Key.Type.DETERMINISTIC_KEY);
-            final Protos.DeterministicKey.Builder detKey = proto.getDeterministicKeyBuilder();
+            final Protos.DeterministicKey.Builder detKey = proto.getDeterministicKey().toBuilder();
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:789]:
```
            if (entries.isEmpty() && isFollowing()) {
                detKey.setIsFollowing(true);
            }
            proto.setDeterministicKey(detKey); //<--- This line is missing below
            if (key.getParent() != null) {
                // HD keys inherit the timestamp of their parent if they have one, so no need to serialize it.
                proto.clearCreationTimestamp();
```

Comment (HashEngineering) [reply core/src/main/java/org/bitcoinj/wallet/DeterministicKeyChain.java:789]: This missing line causes about 11 tests to fail and probably has a lot of other side effects when saving the wallet protobuf file.

File: core/src/main/java/org/bitcoinj/protocols/channels/PaymentChannelClient.java [hunk 2]
```diff
@@ -369,9 +369,9 @@ private void receiveRefund(Protos.TwoWayChannelMessage refundMsg, @Nullable KeyP
             // Make an initial payment of the dust limit, and put it into the message as well. The size of the
             // server-requested dust limit was already sanity checked by this point.
             PaymentChannelClientState.IncrementedPayment payment = state().incrementPaymentBy(Coin.valueOf(minPayment), userKey);
-            Protos.UpdatePayment.Builder initialMsg = contractMsg.getInitialPaymentBuilder();
-            initialMsg.setSignature(ByteString.copyFrom(payment.signature.encodeToBitcoin()));
-            initialMsg.setClientChangeValue(state.getValueRefunded().value);
+//            Protos.UpdatePayment.Builder initialMsg = contractMsg.getInitialPaymentBuilder();
+//            initialMsg.setSignature(ByteString.copyFrom(payment.signature.encodeToBitcoin()));
+//            initialMsg.setClientChangeValue(state.getValueRefunded().value);
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/protocols/channels/PaymentChannelClient.java:374]: I will investigate the reason for these lines being commented as well.

This is not part of the bitcoinj PR that was used as a reference.

Comment (HashEngineering) [reply core/src/main/java/org/bitcoinj/protocols/channels/PaymentChannelClient.java:374]: This section of code and many others related to payment channels was removed from bitcoinj on July 4, 2019.

https://github.com/schildbach/bitcoinj/commit/d67f060e27f38e4117ba20f38d6ae8bd1130ab34

Is any of this related to BIP70?

Comment (HashEngineering) [reply core/src/main/java/org/bitcoinj/protocols/channels/PaymentChannelClient.java:374]: `Protos.UpdatePayment.Builder initialMsg = provideContractBuilder.getInitialPayment().toBuilder();`

The first line of the three can be changed to this.

**Comment by HashEngineering** [comment]
I see one more bug that needs to be looked into which is that the sign (+/-) on the path numbers is negated. 

```
type: DETERMINISTIC_KEY
secret_bytes: "\354B\331\275;\000\254?\3428\006\220G\365\243\333s\260s\213R\313\307\377f\331B\351\327=\001\333"
public_key: "\002\357\\\252\376]\023\315\'\316`\317\362\032@\232\"\360\331\335\221] `\016,\351<\b\300\225\032m"
creation_timestamp: 100000000
deterministic_key {
  chain_code: "\370\017\223\021O?.@gZ|\233j\3437\317q-\241!\177J \323\'\264s\203\314\321\v\346"
  path: 2147483648
}
output_script_type: P2PKH
```


```
# org.bitcoinj.wallet.Protos$Key@f8b44ac5
creation_timestamp: 100000000
deterministic_key {
  chain_code: "\370\017\223\021O?.@gZ|\233j\3437\317q-\241!\177J \323\'\264s\203\314\321\v\346"
  path: -2147483648
}
output_script_type: P2PKH
public_key: "\002\357\\\252\376]\023\315\'\316`\317\362\032@\232\"\360\331\335\221] `\016,\351<\b\300\225\032m"
secret_bytes: "\354B\331\275;\000\254?\3428\006\220G\365\243\333s\260s\213R\313\307\377f\331B\351\327=\001\333"
type: DETERMINISTIC_KEY
```

2147483648 = 0H or 0x80000000 (nonlite)
-2147483648  = 0H or 0x80000000 (lite)

not sure why this is happening

**Comment by HashEngineering** [comment]
Its almost like the non-lite version converts the Int to Long before displaying, but the version doesn't
and displays it as an `int` this is not observed with bitcoinj.

**Comment by tomasz-ludek** [comment]
> this is not observed with bitcoinj.

This can be caused due to the different `protobuf-javalite` lib versions (`3.9.2` in bitcoinj vs `3.11.0` in dashj). Versions prior to `3.11.0` can not be used on Android  prior to API version 26.

**Comment by HashEngineering** [comment]
I tried v3.9.2 and that had the same issue and then discovered that I was wrong about bitcoinj.

The lite versions have a toString() format that differs in the sort order of fields, it has a header line for objects like this `# org.bitcoinj.wallet.Protos$Key@f8b44ac5` which prevents comparisons from succeeding and lastly large integers are rendered as negative numbers.

What I had read earlier about the lite versions not having a `toString()` method was incorrect.  But the implementation is different.

**Pull request merged**